e2e/internal/keyexec/keyexec.go

```diff
@@ -0,0 +1,347 @@
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	expect "github.com/Netflix/go-expect"
+	"github.com/sylabs/singularity/e2e/internal/e2e"
+	"github.com/sylabs/singularity/internal/pkg/test"
+	"github.com/sylabs/singularity/pkg/sypgp"
+)
+
+type testingEnv struct {
+	// base env for running tests
+	CmdPath     string `split_words:"true"`
+	TestDir     string `split_words:"true"`
+	RunDisabled bool   `default:"false"`
+}
+
+var testenv testingEnv
+
+// PullDefaultPublicKey will pull the public Sylabs Admin key
+func PullDefaultPublicKey(t *testing.T) {
+	e2e.LoadEnv(t, &testenv)
+
+	argv := []string{"key", "pull", "F69C21F759C8EA06FD32CCF4536523CE1E109AF3"}
+
+	execKey := exec.Command(testenv.CmdPath, argv...)
+
+	out, err := execKey.CombinedOutput()
+	if err != nil {
+		t.Log(string(out))
+		t.Fatalf("Unable to pull key: %v", err)
+	}
+}
+
+// RemoveDefaultPublicKey will pull the public Sylabs Admin key
+func RemoveDefaultPublicKey(t *testing.T) {
+	e2e.LoadEnv(t, &testenv)
+
+	argv := []string{"key", "remove", "F69C21F759C8EA06FD32CCF4536523CE1E109AF3"}
+	execKey := exec.Command(testenv.CmdPath, argv...)
+
+	out, err := execKey.CombinedOutput()
+	if err != nil {
+		t.Log(string(out))
+		t.Fatalf("Unable to pull key: %v", err)
+	}
+}
+
+func RemovePublicKeyring(t *testing.T) {
+	err := os.Remove(sypgp.PublicPath())
+	if err != nil {
+		t.Fatalf("Unable to remove public keyring: %v", err)
+	}
+}
+
+// RemoveSecretKeyring will delete your secret keyring.
+func RemoveSecretKeyring(t *testing.T) {
+	err := os.Remove(sypgp.SecretPath())
+	if err != nil {
+		t.Fatalf("Unable to remove secret keyring: %v", err)
+	}
+}
+
+func RemoveKeyring(t *testing.T) {
+	err := os.RemoveAll(sypgp.DirPath())
+	if err != nil {
+		t.Fatalf("Unable to remove keyring directory: %v", err)
+	}
+}
+
+// ImportKey will import a key from kpath.
+func ImportKey(t *testing.T, kpath string) (string, []byte, error) {
+	e2e.LoadEnv(t, &testenv)
+
+	argv := []string{"key", "import", kpath}
+	execKey := exec.Command(testenv.CmdPath, argv...)
+
+	cm := fmt.Sprintf("%s %s", testenv.CmdPath, strings.Join(argv, " "))
+
+	b, err := execKey.CombinedOutput()
+
+	return cm, b, err
+}
+
+// ImportPrivateKey will take a private key file (kpath) and import it.
+func ImportPrivateKey(t *testing.T, kpath string) (string, []byte, error) {
+	e2e.LoadEnv(t, &testenv)
+
+	c, err := expect.NewConsole()
+	if err != nil {
+		t.Fatal("Unable to start new console: ", err)
+	}
+	defer c.Close()
+
+	exportCmd := []string{"key", "import", kpath}
+
+	cmd := exec.Command(testenv.CmdPath, exportCmd...)
+	cmd.Stdin = c.Tty()
+
+	buf := bytes.NewBuffer(nil)
+	cmd.Stderr = buf
+	cmd.Stdout = buf
+
+	go func() {
+		c.ExpectEOF()
+	}()
+
+	err = cmd.Start()
+	if err != nil {
+		t.Fatal("Unable to start command: ", err)
+	}
+
+	// Send the passcode to singularity. The first one is the old
+	// password, the next two are the new passowrd.
+	c.Send("e2etests\n")
```

Comment:
the key you are importing will always have this passphrase? Shouldn't this be a parameter to the function?

Reply:
For simplicity its I hard coded it _for now_. Right now its not testing passwords/wrong passwords. Should I change this now?

Reply:
you can place this value (`e2etests`, without the newline) in a variable so that at least there's a hint that this will change / needs to change.

Reply:
Yes, I should of done that before, but it does need the newline. I'll add a comment about that.

Reply:
you can add the new line as part of sending the password using expect.

What I mean is that the password does not contain the newline, so to make it clear that the variable contains the password (and not something else), it should not contain the newline.

Reply:
Got it! okay :+1: